### PR TITLE
Version 1

### PR DESCRIPTION
**Pull request by joe-bell**
> ⚠️ **Please note**: V1 is still a work in progress and prone to bugs/sudden API changes.

# Version 1

## TL;DR

- **raam** V1 introduces a new "mixins" API to roll-your-own layout primitives.
- Support most CSS-in-JS libraries and theoretically beyond React.js (testing in progress)
- 78% smaller bundle size (minified):
  - minified: from 22.3kB to 4.8kB (**78% smaller**)
  - gzipped: from 7.4kB to 1.8kB (**75% smaller**)
- Previous components have moved to `@raam/theme-ui` under a single `Flex` component (with `variant`s).

[**See the updated docs for further usage guidelines**][docs]

## Problem

- **Tied to Theme UI (and React.js)**

  Initially, **raam** was built to meet my own use-case for needing layout primitives across multiple Theme UI projects. Given that there are many options to choose from in the CSS-in-JS world, it would be foolish not to support others.

- **Large bundle size**

  Relying heavily on third-party libraries makes for a rather large bundle size:

  - minified: `22.3kB`
  - gzipped: `7.4kB`

- **Lack of extensibility**

  raam's layout components make the assumption that you want each item to be _wrapped_ with styles, rather than combined with the item's styles.

  Although this approach works well to separate concerns, it makes use-case changes to individual items (such as `flex`, or `flexGrow`) [difficult/near impossible](https://github.com/joe-bell/raam/issues/19).

- **Implementation isn't clear**
  - At the end of the day, it's essentially just a collection of `flex`-based layouts

## Solution

A complete refactor and separation of concerns

- `raam` (mixins)
  
  Introduces a new 'Mixins' API to support (and encourage) rolling-your-own layout primitives, starting with `flexbox()`

  Previous layouts such as 'inline' and 'wrap' are accessible via the `variant` option: `flexbox({ variant: "wrap" })`.

  [See the updated docs for usage guidelines][docs].

  Under-the-hood, raam replaces `gap` with a set of custom properties to control parent and child margins. These are toggled off (back to `initial`) depending on the requirements of specific flex properties; a technique heavily inspired by [David Khourshid's "prop-and-lock" technique](https://twitter.com/davidkpiano/status/1284155737720205313?lang=en).

  > e.g. when `flexWrap` is `nowrap` (for 'stack'-based layouts) the negative offset margin on the flex parent is toggled off. 

- `@raam/theme-ui`

   If rolling-your-own layout components isn't for you, the previous Theme UI components have been merged into a single `<Flexbox />` component, housed in a new package:

   - `<Stack />` ➡ `<Flexbox variant="vStack" />`
   - `<Wrap />` ➡ `<Flexbox variant="wrap" />`
   - `<Inline />` ➡ `<Flexbox variant="inline" />`

### Benefits

Splitting out the Theme UI components and self-composing the core functionality drastically improves `raam`'s bundle size:
  - minified: from 22.3kB to 4.8kB (**78% smaller**)
  - gzipped: from 7.4kB to 1.8kB (**75% smaller**)

More importantly, this new architecture makes it a **lot** easier to maintain; especially around inheriting responsive styles a **ton** easier.

### Caveats

- "Why not just use `:first-child`?"

  Some CSS-in-JS SSR techniques rely on injecting inline `<style>` components which alter the DOM order. Using `:first-child` would cause these to gain a margin. Manually defining `index` is a bit of a pain, but it makes it ensures stability.
- Combining `child` and `parent` causes conflicting styles
- CSS Custom Properties need thorough testing with various libraries before release.

## Future

I'm still keen to implement [`rowGap` and `columnGap`](https://github.com/joe-bell/raam/issues/8) options, and I hope this PR makes it a little easier to add in later.

Another idea is to distribute plain CSS versions of raam's primitives, along with a PostCSS plugin

## Todo

- [x] Initial project architecture
- [x] Migrate flex-gap system to bespoke CSS custom property setup
- [x] Responsive style props
- [ ] Responsive variants
- [x] Extend and re-implement types
  - [x] External `gap` prop (should support number)
  - [x] Internal `flexGap` prop
  - [x] CSS output (**help needed** and appreciated) should support major libraries/React.js style attribute
- [x] Extend and update docs
  - [x] Add mixins with props table
  - [x] Add recipes (for users to build their own layout primitives)
- [ ] Extend and update tests
  - [x] Figure out why I can't run tests in `raam` from `src`. Maybe babel?
- [x] Migrate components to `@raam/theme-ui`
  - [x] Support Theme UI responsive props (perhaps a theme transform function in `raam`)
  - [ ] Docs
  - [ ] Tests
- [x] Add `@raam/react`
  - [ ] Docs
  - [ ] Tests
- [x] Add `@raam/stitches`
  - [ ] Docs
  - [ ] Tests
- [x] Add `@raam/tailwind`
  - [ ] Docs
  - [x] Tests
  - [x] Potentially extract some functionality to `@raam/core`
- [ ] Organise CSS output (vars, then properties, then media queries)
- [ ] Final cleanup/optimisations/design

---

[docs]: https://raam-git-rfc-v1.joebell.vercel.app/

### CODE REVIEW

**Comment by hasparus** [comment]
I'm super excited for this, Joe.

**Comment by joe-bell** [comment]
Thanks @hasparus, that means a lot! I've been meaning to ping you for some advice about these types actually (because you're much more of a TypeScript wizard than I am)

**Comment by joe-bell** [comment]
Just a heads up; I'm working on a [`@raam/stitches` plugin](https://twitter.com/joebell_/status/1314502301021859840) too

Once things are more stable, I'll include in this PR 👀

File: packages/raam/src/index.ts [hunk 1]
```diff
@@ -1,8 +1,267 @@
...
+  const polyfill = {
+    /**
+     * Replace `flexGap` with custom properties.
+     * These are toggled off (via `initial`) by the appropriate flex property
+     * in the latter step.
+     */
+    flexGap: {
+      [CSS_VARS.FLEX_GAP_TOP]: value,
+      [CSS_VARS.FLEX_GAP_LEFT]: value,
+      [CSS_VARS.FLEX_GAP_OFFSET]: `calc(${value} * -0.5)`,
```
Comment (MartijnCuppens) [inline packages/raam/src/index.ts:106]: Using `calc(var(--raam-fg) * -1)` (rendered css) here would make it a little easier to update the gap size size you'll only need to adjust `--raam-fg`

Comment (joe-bell) [reply packages/raam/src/index.ts:106]: Thanks for this! To be honest, I just went for this approach because I had a hunch there were perf concerns when it comes to using `calc` with aliased css custom properties - I can't find anything to back that up though. Maybe something to dig into before changing, just in case

Comment (MartijnCuppens) [reply packages/raam/src/index.ts:106]: fwiw, the `calc()` function is calculated only once, see https://codepen.io/MartijnCuppens/pen/ZEOYxoY?editors=1100
The `--raam-fg` property is defined in the element itself, so I don think there won't be any significant performance issues.

**Comment by afzalsayed96** [comment]
You want a hand for the stitches plugin, @joe-bell?
Can't wait to get my hands on this 😁

**Comment by joe-bell** [comment]
Haha thanks for the offer @afzalsayed96 but I'm all good, it'll be ready soon and you'll be the first to know 🙂

File: packages/raam/src/index.ts (hunk 1 above)
Comment (joe-bell) [reply packages/raam/src/index.ts:91]: I'll dig into this some more, thanks for the suggestion and example

**Comment by joe-bell** [comment]
**Available for testing (but mostly undocumented, so please use with caution)**

- `raam@1.0.0-alpha.33`
- `@raam/stitches@1.0.0-alpha.33`
- `@raam/theme-ui@1.0.0-alpha.33`

cc @afzalsayed96

**Comment by joe-bell** [comment]
🦜 **Latest Releases for Testing**

> ⚠️ Mostly undocumented, so to reiterate; please use with caution

- `raam@1.0.0-alpha.46`
- `@raam/react@1.0.0-alpha.46` 🆕 
- `@raam/stitches@1.0.0-alpha.46` (`withTheme` now `useTheme` for consistency)
- `@raam/tailwind@1.0.0-alpha.46` 🆕 
- `@raam/theme-ui@1.0.0-alpha.46`

**Comment by joe-bell** [comment]
Disclaimer: This PR isn't dead, it's just not a priority 😅

**Comment by afzalsayed96** [comment]
Hey @joe-bell,
Can you point us to where this docs site is?

image

**Comment by joe-bell** [comment]
Hey @afzalsayed96, the docs don't exist yet but you can find the CodeSandbox example here:
https://codesandbox.io/s/raam-x-stitches-rcycb?file=/pages/index.tsx

Just to reiterate once more, this is super WIP so please use with caution

**Comment by joe-bell** [comment]
Resolving some git spaghetti thanks to @renovate-bot 😂 
This is not merged
